### ISSUE TITLE
PageRequestDelegateFactory is relying on "null" injection for optional service IActionContextAccessor (preview.5.21301.17)

### ISSUE THREAD

**Issue by lord-executor** [dotnet/aspnetcore#34335]
### Describe the bug
As stated in the title, the [PageRequestDelegateFactory](https://source.dot.net/#Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/PageRequestDelegateFactory.cs,ee4763b5a0192f44) is relying on a `null` value being injected into its constructor if the dependency of `IActionContextAccessor` is not configured in the DI container which seems to be the case in all default scenarios. This makes it unnecessarily difficult to replace the built-in DI container with a different one like for example Ninject.

Implicitly injecting null into a service through dependency injection is generally considered to be a bad practice and this is the first and only case of doing this in ASP.NET Core that I have discovered so far.

In [ControllerRequestDelegateFactory](https://source.dot.net/#Microsoft.AspNetCore.Mvc.Core/Routing/ControllerRequestDelegateFactory.cs,ae57ea7cdacba71c) the same basic problem of the same optional dependency has been solved by introducing a separate overload that does not require the `IActionContextAccessor` service which seems to be a much more "DI friendly" way of dealing with that.

### To Reproduce
To reproduce the issue, please use https://github.com/lord-executor/AspNetCoreNullInjectRepro which is using `Ninject.Web.AspNetCore` to replace the built-in DI container with Ninject.

### Exceptions (if any)
Of course the exception is caused by Ninject not being able to find an implementation of `IActionContextAccessor` which is not the _real_ problem here, but it helps to illustrate the problem.
```
fail: Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware[1]
      An unhandled exception has occurred while executing the request.
      Ninject.ActivationException: Error activating IActionContextAccessor using conditional implicit self-binding of IActionContextAccessor
      Provider returned null.
      Activation path:
        6) Injection of dependency IActionContextAccessor into parameter actionContextAccessor of constructor of type PageRequestDelegateFactory
        5) Injection of dependency IRequestDelegateFactory into parameter requestDelegateFactories of constructor of type ActionEndpointFactory
        4) Injection of dependency ActionEndpointFactory into parameter endpointFactory of constructor of type DefaultPageLoader
        3) Injection of dependency PageLoader into parameter loader of constructor of type DynamicPageEndpointMatcherPolicy
        2) Injection of dependency MatcherPolicy into parameter policies of constructor of type DfaMatcherBuilder
        1) Request for DfaMatcherBuilder

      Suggestions:
        1) Ensure that the provider handles creation requests properly.

         at Ninject.Activation.Context.ResolveInternal(Object scope) in Ninject.dll:token 0x6000374+0x76
         at Ninject.Activation.Context.Resolve() in Ninject.dll:token 0x6000373+0x8b
         at Ninject.KernelBase.Resolve(IRequest request, Boolean handleMissingBindings) in Ninject.dll:token 0x6000062+0x26c
         at Ninject.KernelBase.Resolve(IRequest request, Boolean handleMissingBindings) in Ninject.dll:token 0x6000062+0x192
         at Ninject.KernelBase.Resolve(IRequest request) in Ninject.dll:token 0x6000059+0x0
         at Ninject.Planning.Targets.Target`1.ResolveWithin(IContext parent) in Ninject.dll:token 0x6000179+0x26
         at Ninject.Activation.Providers.StandardProvider.GetValue(IContext context, ITarget target) in Ninject.dll:token 0x60003f2+0x59
         at Ninject.Activation.Providers.StandardProvider.<>c__DisplayClass15_0.<Create>b__0(ITarget target) in Ninject.dll:token 0x6000533+0x0
         at System.Linq.Enumerable.SelectArrayIterator`2.ToArray() in System.Linq.dll:token 0x60001c3+0x12
         at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source) in System.Linq.dll:token 0x600011a+0x1b
         at Ninject.Activation.Providers.StandardProvider.Create(IContext context) in Ninject.dll:token 0x60003f1+0x7c
         at Ninject.Activation.Context.ResolveInternal(Object scope) in Ninject.dll:token 0x6000374+0x2d
         at Ninject.Activation.Context.Resolve() in Ninject.dll:token 0x6000373+0x72
         at Ninject.KernelBase.<>c__DisplayClass40_0.<Resolve>b__6(IBinding binding) in Ninject.dll:token 0x6000450+0x0
         at System.Linq.Enumerable.WhereSelectListIterator`2.MoveNext() in System.Linq.dll:token 0x600024b+0x45
         at System.Linq.Enumerable.CastIterator[TResult](IEnumerable source)+MoveNext() in System.Linq.dll:token 0x600005f+0x5f
         at System.Collections.Generic.LargeArrayBuilder`1.AddRange(IEnumerable`1 items) in System.Linq.dll:token 0x6000014+0x3d
         at System.Collections.Generic.EnumerableHelpers.ToArray[T](IEnumerable`1 source) in System.Linq.dll:token 0x6000029+0x33
         at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source) in System.Linq.dll:token 0x600011a+0x11
         at Microsoft.AspNetCore.Mvc.Routing.ActionEndpointFactory..ctor(RoutePatternTransformer routePatternTransformer, IEnumerable`1 requestDelegateFactories) in Microsoft.AspNetCore.Mvc.Core.dll:token 0x6000511+0x26
         at DynamicInjector9a145b84497a4c95995ad2ca11fdc9d4(Object[] )
         at Ninject.Activation.Providers.StandardProvider.Create(IContext context) in Ninject.dll:token 0x60003f1+0xa3
         at Ninject.Activation.Context.ResolveInternal(Object scope) in Ninject.dll:token 0x6000374+0x2d
         at Ninject.Activation.Context.Resolve() in Ninject.dll:token 0x6000373+0x72
         at Ninject.KernelBase.Resolve(IRequest request, Boolean handleMissingBindings) in Ninject.dll:token 0x6000062+0x26c
         at Ninject.KernelBase.Resolve(IRequest request) in Ninject.dll:token 0x6000059+0x0
         at Ninject.Planning.Targets.Target`1.ResolveWithin(IContext parent) in Ninject.dll:token 0x6000179+0x26
         at Ninject.Activation.Providers.StandardProvider.GetValue(IContext context, ITarget target) in Ninject.dll:token 0x60003f2+0x59
         at Ninject.Activation.Providers.StandardProvider.<>c__DisplayClass15_0.<Create>b__0(ITarget target) in Ninject.dll:token 0x6000533+0x0
         at System.Linq.Enumerable.SelectArrayIterator`2.ToArray() in System.Linq.dll:token 0x60001c3+0x12
         at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source) in System.Linq.dll:token 0x600011a+0x1b
         at Ninject.Activation.Providers.StandardProvider.Create(IContext context) in Ninject.dll:token 0x60003f1+0x7c
         at Ninject.Activation.Context.ResolveInternal(Object scope) in Ninject.dll:token 0x6000374+0x2d
         at Ninject.Activation.Context.Resolve() in Ninject.dll:token 0x6000373+0x72
         at Ninject.KernelBase.Resolve(IRequest request, Boolean handleMissingBindings) in Ninject.dll:token 0x6000062+0x26c
         at Ninject.KernelBase.Resolve(IRequest request) in Ninject.dll:token 0x6000059+0x0
         at Ninject.Planning.Targets.Target`1.ResolveWithin(IContext parent) in Ninject.dll:token 0x6000179+0x26
         at Ninject.Activation.Providers.StandardProvider.GetValue(IContext context, ITarget target) in Ninject.dll:token 0x60003f2+0x59
         at Ninject.Activation.Providers.StandardProvider.<>c__DisplayClass15_0.<Create>b__0(ITarget target) in Ninject.dll:token 0x6000533+0x0
         at System.Linq.Enumerable.SelectArrayIterator`2.ToArray() in System.Linq.dll:token 0x60001c3+0x12
         at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source) in System.Linq.dll:token 0x600011a+0x1b
         at Ninject.Activation.Providers.StandardProvider.Create(IContext context) in Ninject.dll:token 0x60003f1+0x7c
         at Ninject.Activation.Context.ResolveInternal(Object scope) in Ninject.dll:token 0x6000374+0x2d
         at Ninject.Activation.Context.Resolve() in Ninject.dll:token 0x6000373+0x72
         at Ninject.KernelBase.<>c__DisplayClass40_0.<Resolve>b__6(IBinding binding) in Ninject.dll:token 0x6000450+0x0
         at System.Linq.Enumerable.WhereSelectListIterator`2.MoveNext() in System.Linq.dll:token 0x600024b+0x45
         at System.Linq.Enumerable.CastIterator[TResult](IEnumerable source)+MoveNext() in System.Linq.dll:token 0x600005f+0x5f
         at System.Collections.Generic.EnumerableHelpers.ToArray[T](IEnumerable`1 source, Int32& length) in System.Linq.dll:token 0x6000026+0xa1
         at System.Linq.Buffer`1..ctor(IEnumerable`1 source) in System.Linq.dll:token 0x60003c2+0x22
         at System.Linq.OrderedEnumerable`1.GetEnumerator()+MoveNext() in System.Linq.dll:token 0x6000391+0x1e
         at Microsoft.AspNetCore.Routing.Matching.DfaMatcherBuilder.ExtractPolicies(IEnumerable`1 policies) in Microsoft.AspNetCore.Routing.dll:token 0x600034c+0x5c
         at Microsoft.AspNetCore.Routing.Matching.DfaMatcherBuilder..ctor(ILoggerFactory loggerFactory, ParameterPolicyFactory parameterPolicyFactory, EndpointSelector selector, IEnumerable`1 policies) in Microsoft.AspNetCore.Routing.dll:token 0x600033c+0x5f
         at DynamicInjectora990da55051c441eb2aac30e340e1b05(Object[] )
         at Ninject.Activation.Providers.StandardProvider.Create(IContext context) in Ninject.dll:token 0x60003f1+0xa3
         at Ninject.Activation.Context.ResolveInternal(Object scope) in Ninject.dll:token 0x6000374+0x2d
         at Ninject.Activation.Context.Resolve() in Ninject.dll:token 0x6000373+0x8b
         at Ninject.KernelBase.Resolve(IRequest request, Boolean handleMissingBindings) in Ninject.dll:token 0x6000062+0x26c
         at Ninject.KernelBase.Resolve(IRequest request) in Ninject.dll:token 0x6000059+0x0
         at Ninject.ResolutionExtensions.GetResolutionIterator(IResolutionRoot root, Type service, Func`2 constraint, IEnumerable`1 parameters, Boolean isOptional, Boolean isUnique) in Ninject.dll:token 0x60000a1+0x2f
         at Ninject.ResolutionExtensions.Get(IResolutionRoot root, Type service, IParameter[] parameters) in Ninject.dll:token 0x6000091+0x0
         at Ninject.Web.AspNetCore.NinjectServiceProvider.GetRequiredService(Type serviceType) in Ninject.Web.AspNetCore.dll:token 0x6000015+0x0
         at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType) in Microsoft.Extensions.DependencyInjection.Abstractions.dll:token 0x600008d+0x2c
         at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider) in Microsoft.Extensions.DependencyInjection.Abstractions.dll:token 0x600008e+0x0
         at Microsoft.AspNetCore.Routing.Matching.DfaMatcherFactory.<CreateMatcher>b__2_0() in Microsoft.AspNetCore.Routing.dll:token 0x6000351+0x0
         at Microsoft.AspNetCore.Routing.Matching.DataSourceDependentMatcher.CreateMatcher(IReadOnlyList`1 endpoints) in Microsoft.AspNetCore.Routing.dll:token 0x600032e+0x0
         at Microsoft.AspNetCore.Routing.DataSourceDependentCache`1.Initialize() in Microsoft.AspNetCore.Routing.dll:token 0x6000077+0x1d
         at System.Threading.LazyInitializer.EnsureInitializedCore[T](T& target, Boolean& initialized, Object& syncLock, Func`1 valueFactory) in System.Private.CoreLib.dll:token 0x60029e3+0x19
         at System.Threading.LazyInitializer.EnsureInitialized[T](T& target, Boolean& initialized, Object& syncLock, Func`1 valueFactory) in System.Private.CoreLib.dll:token 0x60029e2+0xf
         at Microsoft.AspNetCore.Routing.Matching.DataSourceDependentMatcher..ctor(EndpointDataSource dataSource, Lifetime lifetime, Func`1 matcherBuilderFactory) in Microsoft.AspNetCore.Routing.dll:token 0x600032b+0x25
         at Microsoft.AspNetCore.Routing.Matching.DfaMatcherFactory.CreateMatcher(EndpointDataSource dataSource) in Microsoft.AspNetCore.Routing.dll:token 0x6000350+0x28
         at Microsoft.AspNetCore.Routing.EndpointRoutingMiddleware.InitializeCoreAsync() in Microsoft.AspNetCore.Routing.dll:token 0x60000b7+0x21
      --- End of stack trace from previous location ---
         at Microsoft.AspNetCore.Routing.EndpointRoutingMiddleware.<Invoke>g__AwaitMatcher|8_0(EndpointRoutingMiddleware middleware, HttpContext httpContext, Task`1 matcherTask) in Microsoft.AspNetCore.Routing.dll:token 0x60000b8+0x6d
         at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context) in Microsoft.AspNetCore.Diagnostics.dll:token 0x60000aa+0x82
```

### Further technical details
- ASP.NET Core version: 6.0.0-preview.5.21301.17
- Output of `dotnet --info`
```
.NET SDK (reflecting any global.json):
 Version:   6.0.100-preview.5.21302.13
 Commit:    d6380bcae7

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19041
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\6.0.100-preview.5.21302.13\

Host (useful for support):
  Version: 6.0.0-preview.5.21301.5
  Commit:  ec3e0b276b

.NET SDKs installed:
  1.0.3 [C:\Program Files\dotnet\sdk]
  1.1.0 [C:\Program Files\dotnet\sdk]
  2.0.0 [C:\Program Files\dotnet\sdk]
  2.1.200 [C:\Program Files\dotnet\sdk]
  2.1.201 [C:\Program Files\dotnet\sdk]
  2.1.202 [C:\Program Files\dotnet\sdk]
  2.1.301 [C:\Program Files\dotnet\sdk]
  2.1.402 [C:\Program Files\dotnet\sdk]
  2.1.524 [C:\Program Files\dotnet\sdk]
  2.1.617 [C:\Program Files\dotnet\sdk]
  2.1.700 [C:\Program Files\dotnet\sdk]
  2.2.300 [C:\Program Files\dotnet\sdk]
  3.1.410 [C:\Program Files\dotnet\sdk]
  5.0.104 [C:\Program Files\dotnet\sdk]
  5.0.203 [C:\Program Files\dotnet\sdk]
  5.0.204 [C:\Program Files\dotnet\sdk]
  6.0.100-preview.5.21302.13 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.All 2.1.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.4 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.24 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.28 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.4 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.24 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.28 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.15 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.16 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.4 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.0-preview.5.21301.17 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 1.0.4 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 1.0.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 1.1.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 1.1.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.4 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.11 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.24 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.28 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.15 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.16 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.4 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.0-preview.5.21301.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.15 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.16 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.4 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.6 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.7 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.0-preview.5.21301.4 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

To install additional .NET runtimes or SDKs:
  https://aka.ms/dotnet-download
```
- Using Microsoft Visual Studio Community 2022 Preview (64-bit) Version 17.0.0 Preview 1.1

**Comment by lord-executor** [comment]
This also seems to relate to #28162 since `IActionContextAccessor` seems to be mostly obsolete by now, but as long as it is still in there and DI injectable, the described problem will persist.

**Comment by pranavkm** [comment]
FYI @davidfowl

**Comment by davidfowl** [comment]
This pattern is expected to work if the container is to be compatible with ASP.NET Core. Last I checked, ninject wasn't compatible. We use this pattern all over the stack. 

Does Ninject.Web.AspNetCore pass the DI conformance tests?

**Comment by pranavkm** [comment]
@davidfowl it feels trivial to add an additional ctor if it fixes the issue. @lord-executor we'd be happy to accept a PR for this change if you'd like to send one

**Comment by davidfowl** [comment]
That seems fine but I'm like 84% sure this pattern exists in other parts of the stack.

**Comment by lord-executor** [comment]
I'll have a look at getting aspnetcore to build locally and run the tests today and getting a small PR started.

But I also have some questions @davidfowl :
* What are the "DI conformance tests" you were referring to? Given that I don't know them we can be pretty sure that Ninject.Web.AspNetCore doesn't pass them and that is something I would like to look into.
* On net5.0 I have several applications running ASP.NET Core MVC, APIs and even Blazor that are working just fine with Ninject.Web.AspNetCore and so far with net6.0 once I provide a "null object" (a copy of the [NullActionContextAccessor](https://source.dot.net/#Microsoft.AspNetCore.Mvc.Core/Infrastructure/ActionContextAccessor.cs,af78edde4d992bd9,references)) everything seems to be back in working order. So, the question is which pattern do you think exists in other parts of the stack? Optional services in general, or optional services based on constructor arguments with `null` default values?

The problem with Ninject is just that it does not allow null injection by default. This can be configured, but then null injection would be allowed _anywhere_ and that just does not seem like a good idea to me. There are other workarounds available, but for those I would have to do a lot of reflection to _find_ the cases with constructor arguments that have an explicit `null` default argument. If there are different constructor overloads, then Ninject can handle that on its own. I'll also have another look at the core Ninject code to see if it would be possible to extend it to better deal with arguments with default values.

**Comment by davidfowl** [comment]
> So, the question is which pattern do you think exists in other parts of the stack? Optional services in general, or optional services based on constructor arguments with null default values?

> optional services based on constructor arguments with null default values?

This one.

> What are the "DI conformance tests" you were referring to? Given that I don't know them we can be pretty sure that Ninject.Web.AspNetCore doesn't pass them and that is something I would like to look into.

AFAIK ASP.NET Core doesn't work with ninject. Here are the conformance tests https://www.nuget.org/packages/Microsoft.Extensions.DependencyInjection.Specification.Tests/6.0.0-preview.6.21352.12.

**Comment by lord-executor** [comment]
@davidfowl : Thanks a lot for the clarification and the link.

Since I am certainly not going to decline an invitation to contribute to ASP.NET Core in _any_ way, I have created a PR https://github.com/dotnet/aspnetcore/pull/34378 that aligns the handling of the `IActionContextAccessor` in `PageRequestDelegateFactory` with the (IMO better, because more explicit and more "compatible") implementation of `ControllerRequestDelegateFactory`.

You are correct that generally speaking, Ninject does not work with ASP.NET Core. The https://github.com/lord-executor/Ninject.Web.AspNetCore integration was created by a coworker of mine and myself because we had a big project using Ninject that needed to be updated to .NET 5. It is definitely not perfect, but it does _work_ in all the cases that we have encountered so far.

With the conformance tests, I can now actually significantly improve the compatibility which is great! There are currently 8 failing tests, some of them relate to service disposal and I am already working on fixes for those. The remaining 5 failing tests all relate to services with open/closed generics which is definitely a bit more difficult but I can probably improve that as well.

Notably, there is _no_ test that explicitly or implicitly covers the injection of constructor arguments with `null` default values. If that is, as you say, a common use case, then I could have a look at extending those conformance tests with something along those lines?
